keps/sig-api-machinery/2558-field-version-openapi/README.md

```diff
@@ -0,0 +1,414 @@
+the feature gates in a comment but these comments are not parseable either.
+
+**Field version**
+
+Currently, when a new field is added to a stable API version, there
+is no way to expose the version of the new field in OpenAPI.
+One way to identify alpha API fields is by comparing the OpenAPI schema
+generated from an apiserver with all APIs enabled and the one generated
+from an apiserver with only stable and beta APIs enabled.
+However, this process is cumbersome.
+
+Since the OpenAPI definition is the source of truth for API documentation,
+this KEP proposes a way to publish field-level versioning and feature gate
+information via OpenAPI.
+
+As a follow-up, this information can also be used to:
+- update the Kubernetes API reference documentation
+- surface [server-side warnings] in client-go and kubectl when a deprecated
+field is being used.
+
+### Goals
+
+- Publish feature gate associated with an API field for built-in API types
+in OpenAPI to document the mapping of API fields with their feature gates.
+- Document the current prerelease status of an API field by publishing
+the prerelease status of the field and the minimum Kubernetes version for
+the prerelease in OpenAPI.
+- Validate that the feature gate mentioned in the API field comment
+is a valid feature gate.
+
+### Non-Goals
+
+- Support publishing field-level versioning for custom resources in OpenAPI.
+- Publish the default status of a feature gate (enabled/disabled by default)
+in OpenAPI.
+- Allow parsing comment tags on feature gates in [`kube_features.go`].
+- Update the Kubernetes API reference docs to include the field-level
+versioning information (out of scope of this KEP, can be done in a follow up).
+
+## Proposal
+
+This KEP proposes changes to the kube-openapi generator to add a new
+vendor extension field `x-kubernetes-field-version` in OpenAPI.
+This extension specifies the prerelease status of an API field,
+the minimum Kubernetes version for the prerelease and the
+feature gate associated with the field.
+
+This information is derived from comment tags on API fields with the format:
+
+```
+// +k8s:openapi-gen:prerelease=<prerelease-status>,minVersion=<k8s-version>,featureGate=<featuregate-name>
+```
+
+- The `prerelease` tag must be one of `alpha`, `beta`, `stable`
+or `deprecated`.
+- The `minVersion` tag denotes that the field is in the prerelease
+since the specified Kubernetes version. It must map to a Kubernetes
+release (e.g. `v1.20`).
+- The `featureGate` tag must be a valid feature gate as defined in [`kube_features.go`].
+
+The comment tag will be removed when the feature gate for the API field is removed.
+This will remove the `x-kubernetes-field-version` vendor extension for
+the field in OpenAPI.
+
+<<[UNRESOLVED]>>
+
+Open questions from @thockin's proposal:
+
+1) Should the tooling derive the prerelease status by looking up the gate name
+(one place to assert).
+
+2) Should we also tag structs or even new Kinds when they are gated?  This is
```

Comment:
I think we should, even if we don't know what to do with that yet.

Reply:
> I think we should, even if we don't know what to do with that yet.

Done :+1: